### ISSUE TITLE
Better API for creating requests from responses

### ISSUE THREAD

**Issue by kmike** [scrapy/scrapy#1940]
Sometime Request needs information about the response to be sent correctly. There are at least 2 use cases:
1. URL encoding depends on response encoding (see https://github.com/scrapy/scrapy/pull/1923);
2. relative URLs should be resolved based on response.url or its base url (see https://github.com/scrapy/scrapy/issues/548).

I think the current API is not good enough. The most obvious code is not correct:

``` py
for url in response.xpath("//a/@href").extract():
    yield scrapy.Request(url, self.parse)
```

To do that correctly user has to write the following:

``` py
for url in response.xpath("//a/@href").extract():
    yield scrapy.Request(response.urljoin(url), self.parse, encoding=response.encoding)
```

Or this:

``` py
for link in LinkExtractor().extract_links(response):
    yield scrapy.Request(link.url, self.parse, encoding=response.encoding)
```

LinkeExtractor solution has gotchas, e.g. canonicalize_url is called by default and fragments are removed. It means that e.g. Ajax crawlable URLs are not handled (no `escaped_fragment` even if a website supports it); it also makes it harder to use Scrapy with scrapy-splash which can handle fragments.

This all is too easy to get wrong; I think just documenting these gotchas is not good enough for a framework - it should make the easiest way to write something the correct way. IMHO in the API shouldn't require user to instantiate weird objects or pass response encoding:

``` py
for url in response.xpath("//a/@href").extract():
    something.send_request(url, self.parse)
```

This can be implemented if we provide a method on Response to send new requests.

A related use case is `async def` functions or methods (https://github.com/scrapy/scrapy/issues/1144#issuecomment-141843616): it is not possible to yield Requests in `async def` functions, so adding a request should be either a method of `self` or a method of `response` if we want to support `async def` callbacks.

`FormRequest.from_response(response, ...)` can also be written as something like `response.submit(...)`.

**Comment by Digenis** [comment]
More LinkExtractor gotchas: #1941

I think providing a method in the response to prepare requests and urls
can be implemented fast and reliably without breaking compatibility.

The rest, sending the request to the scheduler, sounds like something that requires a separate issue.

**Comment by kmike** [comment]
Another way to fix response.encoding problem is to create a middleware which sets request.encoding to response.encoding if it is not set explicitly. But it would be backwards incompatible, and @dangra didn't like middleware approach for relative URLs and ruled them out (which was a good decision).

**Comment by dangra** [comment]
> A related use case is async def functions or methods (#1144 (comment)): it is not possible to yield Requests in async def functions, so adding a request should be either a method of self or a method of response if we want to support async def callbacks.

this is not true anymore since https://www.python.org/dev/peps/pep-0525/, right?

**Comment by kmike** [comment]
We can also add Selector support to Response.Request, in order to write this:

```python
for a in response.css("a.my-link"):
    yield response.Request(a, self.parse)
```

instead of this:

```python
for href in response.css("a.my-link::attr(href)").extract():
    yield response.Request(href, self.parse)
```

(or this, using scrapy.Request):

```python
for href in response.css("a.my-link::attr(href)").extract():
    yield scrapy.Request(response.urljoin(href), self.parse, encoding=response.encoding)
```

**Comment by kmike** [comment]
@dangra yep! It is Python 3.6+ only, but I'm fine with that.

**Comment by dangra** [comment]
I like the idea of adding response methods to generate compelling requests, but I am not fond of using capitalized method names that also mimics scrapy builtin classes.

IMO we can set the request class the *new methods* are going to use to instantiate requests as Response class attribute.  So for example xmlrpc responses default to `scrapy.http.request.XmlRpcRequest`

**Comment by dangra** [comment]
> IMO we can set the request class the new methods are going to use to instantiate requests as Response class attribute. So for example xmlrpc responses default to scrapy.http.request.XmlRpcRequest

this is food for another discussion thread, my point is that using capitalized method names for the feature described in this ticket is not future compatible.

**Comment by kmike** [comment]
If adding a link to Crawler to the Response is too much (I personally feel it is too much) then we don't have to 'reserve' response.follow / response.submit names. If Response.Request deviates from vanilla scrapy.Request further (e.g. if it gets Selector support) then I think it adds points to response.follow (or alike) names. The same for response.submit vs response.FormRequest; the former gives a chance to cleanup the API. Disadvantages of "sweet" names still apply though.

**Comment by kmike** [comment]
> IMO we can set the request class the new methods are going to use to instantiate requests as Response class attribute. So for example xmlrpc responses default to scrapy.http.request.XmlRpcRequest

I'm not sure I got it correctly; could you please provide an example?

**Comment by dangra** [comment]
> If adding a link to Crawler to the Response is too much (I personally feel it is too much) then we don't have to 'reserve' response.follow / response.submit names.

I feel inclined towards not linking Crawler to responses too.

> If Response.Request deviates from vanilla scrapy.Request further (e.g. if it gets Selector support) then I think it adds points to response.follow (or alike) names. The same for response.submit vs response.FormRequest; the former gives a chance to cleanup the API.

+1

> Disadvantages of "sweet" names still apply though.

not sure what are the disadvantages you refer to :)

**Comment by kmike** [comment]
> not sure what are the disadvantages you refer to :)

As a newcomer, a very easy mistake to make is to write

```python
for a in response.css("a.my-link"):
    response.follow(a, self.parse)
```

instead of

```python
for a in response.css("a.my-link"):
    yield response.follow(a, self.parse)
```

With `yield scrapy.Request` it is clear we're creating a Request object, but not necessarily executing an action. `response.follow` reads like an action, but it does nothing on its own. But I guess it applies to all async APIs, so people can get used to this.

**Comment by kmike** [comment]
Ideally, we should support both `resp2 = await response.follow(url)` and `yield response.follow(url)`

**Comment by kmike** [comment]
So, wishlist for response.follow:

* support relative and absolute URLs;
* set encoding correctly;
* support Selector objects with `a` elements (what about other elements? img src?);
* support Link objects;
* support all Request options.

**Comment by dangra** [comment]
> Ideally, we should support both resp2 = await response.follow(url) and yield response.follow(url)

Yes, I think it is the desired interface and should be possible to implement.

**Comment by Gallaecio** [comment]
Fixed in #2540.
